app/components/autocomplete/channel_mention/channel_mention.tsx

```diff
@@ -0,0 +1,290 @@
+
+const keyExtractor = (item: Channel) => {
+    return item.id;
+};
+
+const getMatchTermForChannelMention = (() => {
+    let lastMatchTerm: string | null = null;
+    let lastValue: string;
+    let lastIsSearch: boolean;
+    return (value: string, isSearch: boolean) => {
+        if (value !== lastValue || isSearch !== lastIsSearch) {
+            const regex = isSearch ? CHANNEL_MENTION_SEARCH_REGEX : CHANNEL_MENTION_REGEX;
+            const match = value.match(regex);
+            lastValue = value;
+            lastIsSearch = isSearch;
+            if (match) {
+                if (isSearch) {
+                    lastMatchTerm = match[1].toLowerCase();
+                } else if (match.index && match.index > 0 && value[match.index - 1] === '~') {
+                    lastMatchTerm = null;
+                } else {
+                    lastMatchTerm = match[2].toLowerCase();
+                }
+            } else {
+                lastMatchTerm = null;
+            }
+        }
+        return lastMatchTerm;
+    };
+})();
+
+const reduceChannelsForSearch = (channels: Channel[], members: MyChannelModel[]) => {
+    return channels.reduce<Channel[][]>(([pubC, priC, dms], c) => {
+        switch (c.type) {
+            case General.OPEN_CHANNEL:
+                if (members.find((m) => c.id === m.id)) {
+                    pubC.push(c);
+                }
+                break;
+            case General.PRIVATE_CHANNEL:
+                priC.push(c);
+                break;
+            case General.DM_CHANNEL:
+            case General.GM_CHANNEL:
+                dms.push(c);
+        }
+        return [pubC, priC, dms];
+    }, [[], [], []]);
+};
+
+const reduceChannelsForAutocomplete = (channels: Channel[], members: MyChannelModel[]) => {
+    return channels.reduce<Channel[][]>(([myC, otherC], c) => {
+        if (members.find((m) => c.id === m.id)) {
+            myC.push(c);
+        } else {
+            otherC.push(c);
+        }
+        return [myC, otherC];
+    }, [[], []]);
+};
+type Props = {
+    cursorPosition: number;
+    isSearch: boolean;
+    maxListHeight: number;
+    myMembers: MyChannelModel[];
+    updateValue: (v: string) => void;
+    onShowingChange: (c: boolean) => void;
+    value: string;
+    nestedScrollEnabled: boolean;
+}
+
+const getStyleFromTheme = makeStyleSheetFromTheme((theme) => {
+    return {
+        listView: {
+            backgroundColor: theme.centerChannelBg,
+            borderRadius: 4,
+        },
+    };
+});
+
+const ChannelMention = ({
+    cursorPosition,
+    isSearch,
+    maxListHeight,
+    myMembers,
+    updateValue,
+    onShowingChange,
+    value,
+    nestedScrollEnabled,
+}: Props) => {
+    const serverUrl = useServerUrl();
+    const theme = useTheme();
+    const style = getStyleFromTheme(theme);
+    const [sections, setSections] = useState<Array<SectionListData<Channel>>>([]);
+    const [channels, setChannels] = useState<Channel[]>([]);
```

Comment:
are we not using the channels that we already have in the database for a little bit of speed? if the user is composing a message on an airplane wouldn't they even use any of the existing channels in the autocomplete?